### DISCUSSION TITLE
License

### DISCUSSION THREAD

**Discussion by drake127** [animetosho/md5-optimisation#4] [Q&A] [answered]
Hi, can you specify what license do you release the source code under?

**Comment by animetosho** [comment, marked as answer]
It's under Public Domain or CC0 (if the former isn't recognised).

**Comment by xtqqczze** [comment]
@animetosho Please consider using an explicit open-source license instead of Public Domain/CC0.

- "Public Domain" isn't itself a software license, and its legal effect varies between jurisdictions.
- CC0 is also problematic for some downstream distributors. In particular, Fedora has had restrictions around CC0 for software, including concerns around the scope of the patent waiver.
- A standard permissive license would make the project much easier for distributions and downstream projects to package and reuse.

Would you consider the same explicit dual-license used by Rust: **MIT OR Apache-2.0**? Rust recommends this combination for maximum ecosystem compatibility, and it provides both the simplicity of MIT and the explicit patent grant of Apache-2.0. ([rust-lang.github.io](https://rust-lang.github.io/api-guidelines/necessities.html))

For example:

> Licensed under either of Apache License, Version or MIT license at your option.

**Reply by animetosho** [reply]
Thanks for the suggestion.  
Feel free to apply whatever license you want, I have no interest in enforcing it.

Although I don't know why you'd want to actually use the code here - it's mostly for demonstration purposes.

**Reply by oech3** [reply]
We want to use this code for uutils/coreutils to achieve perf beter than OpenSSL. uutils/coreutils is MIT.

**Reply by animetosho** [reply]
I believe the main optimisation in the article has [since landed in OpenSSL](https://github.com/openssl/openssl/pull/25737), so you may already be using it.  
I don't believe OpenSSL implements the other tricks listed here, though the benefits of those may not be observable.

The code in this repo is only really meant for the performance graph in the article, not designed to be used directly in an application.  
I do have an application oriented implementation [in ParPar](https://github.com/animetosho/ParPar/tree/master/hasher), but it comes with a bunch of other stuff you likely don't want. Maybe one day I'll extract it into an independent library.

**Reply by xtqqczze** [reply]
Just to be clear, we don't intend to use the code in this repository directly. We want to use https://github.com/ogital-net/fast-md5, which was ported from this project. As a result, it could reasonably be considered a derivative work, so the ambiguity around the original project's licensing is a concern.

For that reason, it would be much safer if this project were explicitly licensed under **MIT OR Apache-2.0**. There is already a PR open to do this: https://github.com/animetosho/md5-optimisation/pull/13.

The concern with relying on CC0 is that it is a copyright/public-domain dedication and does not provide an express patent licence. For software that incorporates or derives from the code, that's an important distinction; we'd prefer a conventional software licence that explicitly grants the relevant patent rights and makes the licensing position unambiguous.

**Reply by animetosho** [reply]
I see, thanks for explaining and putting up with my questions.

**Reply by xtqqczze** [reply]
Thank you for merging https://github.com/animetosho/md5-optimisation/pull/13!

**Comment by oech3** [comment]
Since we don't want to depend on C, using this via OpenSSL is not a solution.
